platforms/posix/src/px4/common/main.cpp

```diff
@@ -572,29 +572,40 @@ void print_usage()
 	printf("        e.g.: px4-commander status\n");
 }
 
-bool is_already_running(int instance)
+bool is_server_running(int instance, bool server)
 {
 	const std::string file_lock_path = std::string(LOCK_FILE_PATH) + '-' + std::to_string(instance);
-	struct flock fl;
 	int fd = open(file_lock_path.c_str(), O_RDWR | O_CREAT, 0666);
 
 	if (fd < 0) {
+		PX4_ERR("is_server_running: failed to create lock file: %s, reason=%s", file_lock_path.c_str(), strerror(errno));
 		return false;
 	}
 
-	fl.l_type   = F_WRLCK;
-	fl.l_whence = SEEK_SET;
-	fl.l_start  = 0;
-	fl.l_len    = 0;
-	fl.l_pid    = getpid();
+	bool result = false;
+
+	// Server is running if the file is locked.
+	// This is true if the non-blocking flock returns EWOULDBLOCK.
+	if (flock(fd, LOCK_EX | LOCK_NB) < 0) {
+		if (errno == EWOULDBLOCK) {
+			result = true;
+
+		} else {
+			PX4_ERR("is_server_running: failed to get lock on file: %s, reason=%s", file_lock_path.c_str(), strerror(errno));
+			result = false;
+		}
+	}
 
-	if (fcntl(fd, F_SETLK, &fl) == -1) {
-		// We failed to create a file lock, must be already locked.
-		return errno == EACCES || errno == EAGAIN;
+	if (result || !server) {
+		// another server is running!
```

Comment:
This is not necessarily true at this point (it could be a client that took the lock).
I'd just move the comment below up here, and remove this one.

Reply:
Good catch, thanks.  It is fixed.